### PR TITLE
FM2-608: Add EpisodeOfCare FHIR2 Module with Mapping of Patient Program

### PR DESCRIPTION
**Pull request by k4pran**
## Description of what I changed

Added support for fhir EpisodeOfCare with a mapping from the openmrs PatientProgram. This is a one-way mapping from the openmrs PatientProgram to EpisodeOfCare and doesn't include support for deleting / creating / updating.

## Issue I worked on

see https://issues.openmrs.org/browse/FM2-608

## Checklist: I completed these to help reviewers :)

- [X] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [X] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [X] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [X] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [X] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: omod/src/main/java/org/openmrs/module/fhir2/web/servlet/FhirRestServlet.java [hunk 1]
```diff
@@ -63,6 +64,9 @@ public class FhirRestServlet extends RestfulServer implements ModuleLifecycleLis
 	private static final List<String> DEFAULT_NARRATIVE_FILES = Arrays.asList(FhirConstants.OPENMRS_NARRATIVES_PROPERTY_FILE,
 	    FhirConstants.HAPI_NARRATIVES_PROPERTY_FILE);
 	
+	@Autowired
+	private EpisodeOfCareFhirResourceProvider dao;
```
Comment (ibacher) [inline omod/src/main/java/org/openmrs/module/fhir2/web/servlet/FhirRestServlet.java:68]: You shouldn't need to autowire this here. (Also the name of the variable is slightly confusing.

Comment (k4pran) [reply omod/src/main/java/org/openmrs/module/fhir2/web/servlet/FhirRestServlet.java]: Removed, this was an artifact left over when I was debugging something

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirEpisodeOfCareServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	@Autowired
+	private FhirEpisodeOfCareDao dao;
+	
+	@Autowired
+	private EpisodeOfCareTranslator translator;
+	
+	@Override
+	protected FhirDao<PatientProgram> getDao() {
+		return dao;
+	}
```
[the remark this answers is not in the document]
Comment (k4pran) [reply api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirEpisodeOfCareServiceImpl.java]: Good catch, removed

File: api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/EpisodeOfCareTranslatorImpl.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+	public PatientProgram toOpenmrsType(@Nonnull EpisodeOfCare episodeOfCare) {
+		notNull(episodeOfCare, "The EpisodeOfCare object should not be null");
+		return this.toOpenmrsType(new PatientProgram(), episodeOfCare);
+	}
+	
+	@Override
+	public PatientProgram toOpenmrsType(@Nonnull PatientProgram patientProgram, @Nonnull EpisodeOfCare episodeOfCare) {
+		log.warn("Translation from FHIR resource EpisodeOfCare to OpenMRS object PatientProgram is not supported.");
+		return null;
+	}
```
[the remark this answers is not in the document]
Comment (k4pran) [reply api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/EpisodeOfCareTranslatorImpl.java]: Done and updated corresponding test

**Comment by slubwama** [comment]
@k4pran  @ibacher how far did this go? It turns out that we need this implementation asap given that we have use of episodeOfCare in our impllemataion

**Comment by k4pran** [comment]
> @k4pran @ibacher how far did this go? It turns out that we need this implementation, given that we have use of episodeOfCare in our implementation

I have addressed outstanding comments, so once @ibacher is happy with the changes it should be ready

**Pull request merged**